nomad/node_endpoint.go

```diff
@@ -211,6 +207,36 @@ func (n *Node) Register(args *structs.NodeRegisterRequest, reply *structs.NodeUp
 	return nil
 }
 
+// shouldCreateNodeEval returns true if the node update may result into
+// allocation updates, so the node should be re-evaluating.
+//
+// Such cases might be:
+// * node health/drain status changes that may result into alloc rescheduling
+// * node drivers or attributes changing that may cause system job placement changes
+func shouldCreateNodeEval(original, updated *structs.Node) bool {
+	if structs.ShouldDrainNode(updated.Status) {
+		return true
+	}
+
+	if original == nil {
+		return transitionedToReady(updated.Status, structs.NodeStatusInit)
+	}
+
+	if transitionedToReady(updated.Status, original.Status) {
+		return true
+	}
+
+	// check fields used by the feasability checkers, through direct
+	// or interpolated constraints.
+	return !(original.ID == updated.ID &&
+		original.Datacenter == updated.Datacenter &&
+		original.Name == updated.Name &&
+		original.NodeClass == updated.NodeClass &&
+		reflect.DeepEqual(original.Attributes, updated.Attributes) &&
+		reflect.DeepEqual(original.Meta, updated.Meta) &&
+		reflect.DeepEqual(original.Drivers, updated.Drivers))
```

Comment:
Hm, some other fields that might need to be considered:
- NodeResources 
- ReservedResources
- HostVolumes

It might also be worth documenting that DrainStrategy and SchedulingEligibility do *not* need to be considered here since they're only mutated through other endpoints that will emit node evals of their own.

Reply:
Good catch on HostVolumes!.

Resource constraints are handled differently: they are handled by blocking evals that got addressed in https://github.com/hashicorp/nomad/pull/5900 .